packages/libs/error-stack/src/fmt/mod.rs

```diff
@@ -885,66 +732,101 @@ impl Opaque {
     }
 }
 
-fn debug_attachments(
-    loc: Option<Line>,
-    position: Position,
+fn debug_attachments_order(emits: impl Iterator<Item = Emit>) -> Vec<String> {
+    let mut out = Vec::with_capacity(emits.size_hint().0);
+    let mut defer = vec![];
+
+    for emit in emits {
+        match emit {
+            Emit::Defer(value) => {
+                defer.push(value);
+            }
+            Emit::Immediate(value) => {
+                out.push(value);
+            }
+        }
+    }
+
+    out.append(&mut defer);
+
+    out
+}
+
+fn debug_attachments_invoke(
     frames: Vec<&Frame>,
     #[cfg(feature = "std")] ctx: &mut HookContextImpl,
-) -> Lines {
-    let last = matches!(position, Position::Final);
+) -> (Opaque, Vec<String>, Vec<String>) {
     let mut opaque = Opaque::new();
 
-    // evaluate all frames to their respective values, will call all hooks with the current context
-    let (next, defer): (Vec<_>, _) = frames
+    let (emits, snippets): (Vec<_>, Vec<_>) = frames
         .into_iter()
         .map(|frame| match frame.kind() {
             #[cfg(feature = "std")]
             FrameKind::Attachment(AttachmentKind::Opaque(_)) | FrameKind::Context(_) => {
-                Report::get_debug_format_hook(|hooks| hooks.call(frame, &mut ctx.as_hook_context()))
+                let mut ctx = ctx.as_hook_context();
+                Report::get_debug_format_hook(|hooks| hooks.call(frame, &mut ctx));
+                ctx.into_parts()
             }
             #[cfg(not(feature = "std"))]
             FrameKind::Attachment(AttachmentKind::Opaque(_)) | FrameKind::Context(_) => {
-                vec![]
+                (vec![], vec![])
             }
             FrameKind::Attachment(AttachmentKind::Printable(attachment)) => {
-                vec![Emit::Next(attachment.to_string())]
+                (vec![Emit::Immediate(attachment.to_string())], vec![])
             }
         })
         .enumerate()
-        .flat_map(|(idx, value)| {
+        .map(|(idx, (emits, snippets))| {
             // increase the opaque counter, if we're unable to determine the actual value of the
             // frame
-            if idx > 0 && value.is_empty() {
+            if idx > 0 && emits.is_empty() {
```

Comment:
I think it's a matter of taste, but do we also check, if `snippets.is_empty()` as well? Having a snippet at least means that the hook was called.

Reply:
Well, this is used for the "opaque" counter. If something returned a snippet, it would not show up in the tree, meaning it is still opaque 🤔

Reply:
I know what this is for, I was just wondering if we still want to count it as opaque if it emitted a snippet.

Reply:
I'd prefer to keep it like it is right now, as the opaque attachments counter does not say a hook wasn't called, but rather that it didn't show up. 

(That's at least how it works in my mind)

but that isn't a strong preference, so I'd be happy to change it if you prefer it the other way.

Reply:
Cc @hashintel/error-stack?